src/verify.c

```diff
@@ -0,0 +1,131 @@
+                   struct Error **err) {
+  // We want to clear the OpenSSL Error queue so that we know when we're in the
+  // cleanup section, any errors we hit are the result of this invocation
+  ERR_clear_error();
+  return_t rv;
+
+  // First, we need to convert from C-strings into BIOs backed by memory
+  BIO *bio_pubkey = BIO_new_mem_buf(pubkey, pubkey_l);
+  BIO *bio_document = BIO_new_mem_buf(document, document_l);
+  BIO *bio_pkcs7 = BIO_new_mem_buf(pkcs7, pkcs7_l);
+
+  // We don't want the BIO_close method here to free() the memory passed in
+  // BIO_set_close(bio_pubkey, BIO_NOCLOSE);
+  // BIO_set_close(bio_document, BIO_NOCLOSE);
+  // BIO_set_close(bio_pkcs7, BIO_NOCLOSE);
+
+  // The data structures needed for verification
+  PKCS7 *p7;
+  X509_STORE *store = NULL;
+  STACK_OF(X509) *certs = NULL;
+  X509 *cert = NULL;
+
+  // I'd prefer to pass an out parameter, but it seems that the second time
+  // that this function executes, we get a segfault in the underlying library.
+  // I'm not sure how best to debug this, but using the pointer returning
+  // option of the function works well
+  //
+  //   if(!PEM_read_bio_PKCS7(bio_pkcs7, &p7, NULL, NULL)) {
+  //
+  p7 = PEM_read_bio_PKCS7(bio_pkcs7, NULL, NULL, NULL);
+  if (p7 == NULL) {
+    rv = VF_EXCEPTION;
+    goto end;
+  }
+
+  store = X509_STORE_new();
+  if (store == NULL) {
+    rv = VF_EXCEPTION;
+    goto end;
+  }
+
+  certs = sk_X509_new_null();
+  if (certs == NULL) {
+    rv = VF_EXCEPTION;
+    goto end;
+  }
+
+  // Same as reading the PKCS#7 file above
+  //   if (!PEM_read_bio_X509(bio_pubkey, &cert, 0, NULL)) {
+  cert = PEM_read_bio_X509(bio_pubkey, NULL, NULL, NULL);
+  if (cert == NULL) {
+    rv = VF_EXCEPTION;
+    goto end;
+  }
+
+  if (0 == sk_X509_push(certs, cert)) {
+    rv = VF_EXCEPTION;
+    goto end;
+  }
+
+  if (1 == PKCS7_verify(p7, certs, store, bio_document, NULL,
+                        PKCS7_NOINTERN | PKCS7_NOVERIFY)) {
+    rv = VF_SUCCESS;
+  } else {
+    rv = VF_FAIL;
+  }
+
+end:
+  if (!BIO_free(bio_document) || !BIO_free(bio_pubkey) ||
+      !BIO_free(bio_pkcs7)) {
+    rv = VF_EXCEPTION;
+  }
+
+  PKCS7_free(p7);
+  X509_STORE_free(store);
+  sk_X509_free(certs);
+  X509_free(cert);
+
+  struct Error *head = NULL;
+  // We're going to send back the last error message so that we can throw
+  // an exception, but only if we've got somewhere to put the error linked
+  // list
+  if (err && rv == VF_EXCEPTION) {
+    struct Error newError;
+
+    unsigned long errorNum;
+
+    do {
+      errorNum = ERR_get_error_line(&newError.file_string, &(newError.line));
+
+      if (!errorNum) {
+        break;
+      }
+
+      newError.reason_string = ERR_reason_error_string(errorNum);
+      newError.lib_string = ERR_lib_error_string(errorNum);
+      newError.func_string = ERR_func_error_string(errorNum);
+      newError.next = head;
+      head = &newError;
+    } while (errorNum);
+  }
```

Comment:
Rather than have `VF_verify` return a list of errors (only one of which is displayed), it would probably be better to use this function to concatenate together a string containing all of the errors in some vaguely human-readable fashion ("vaguely" because I've seen OpenSSL's errors before!).  Then the error output parameter from `VF_verify` would be a string with the responsibility of freeing that string on the caller.  As-is, it looks like the error list is leaked.

Reply:
Yes, it would be better, but I think that for the time being, this is enough.  Per the discussion above about how the memory for the Error list is allocated, I think that a bit of a rethink is needed here around how errors are handled.